File: app/src/main/java/com/tkamat/android/youwatch/TopicService.java
package com.tkamat.android.youwatch;

import android.app.*;
import android.app.job.JobParameters;
import android.app.job.JobService;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.Uri;
import android.os.Build;
import android.support.annotation.IntDef;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.NotificationManagerCompat;
import android.util.Log;
import android.widget.Toast;
import com.google.api.services.youtube.model.Video;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Tushaar Kamat
 * @version 9/26/17
 */

public class TopicService extends JobService {
    public static final String TAG = "topic_service";
    public static final String CHANNEL_ID = "topic_channel";

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    @Override
    public boolean onStartJob(JobParameters jobParameters) {
        if (!isNetworkAvaibaleAndConnected()) {
            Util.scheduleJob(this);
            return false;
        }
        List<Topic> topics = TopicList.get(this).getTopics();
        for (Topic t : topics) {
            if (t.ismEnabled()) {
                List<String> oldVideoIDs = t.getmTopicSearcher().getmVideoIDs();
                t.setmTopicSearcher(new TopicSearcher(t));
                t.getmTopicSearcher().searchForIDs().searchForVideos();
                List<String> newVideoIDs = t.getmTopicSearcher().getmVideoIDs();
                List<Video> newVideoResults = t.getmTopicSearcher().getmResults();
                List<String> uniqueVideoIDs = new ArrayList<>();
                for (int i = 0; i < newVideoIDs.size(); i++) {
                    if (newVideoResults.get(i) != null && !oldVideoIDs.contains(newVideoIDs.get(i)) && !t.getmNotifiedVideos().contains(newVideoIDs.get(i))) {
                        uniqueVideoIDs.add(newVideoIDs.get(i));
                        String title = "New From " + newVideoResults.get(i).getSnippet().getChannelTitle();
                        String body = newVideoResults.get(i).getSnippet().getTitle();
                        t.getmNotifiedVideos().add(newVideoIDs.get(i));
                        TopicList.get(this).updateTopic(t);
                        Util.createNotification(newVideoIDs.get(i), title, body, this);
                    }
                }
                Log.i(TAG, "Topic refreshed");
                Log.i(TAG, uniqueVideoIDs.toString());
            }
        }
        Util.scheduleJob(this);
        return true;
    }

    @Override
    public boolean onStopJob(JobParameters jobParameters) {
        return true;
    }

    @Override
    public void onDestroy() {
    }

    public static Intent newIntent (Context context) {
        return new Intent(context, TopicService.class);
    }

    private boolean isNetworkAvaibaleAndConnected() {
        ConnectivityManager cm = (ConnectivityManager) getSystemService(CONNECTIVITY_SERVICE);
        boolean isNetworkAvailable = cm.getActiveNetworkInfo() != null;
        boolean isNetworkConnected = isNetworkAvailable && cm.getActiveNetworkInfo().isConnected();

        return isNetworkConnected;
    }

}

File: app/src/main/java/com/tkamat/android/youwatch/TopicService.kt
package com.tkamat.android.youwatch

import android.app.*
import android.app.job.JobParameters
import android.app.job.JobService
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.util.Log

import java.util.ArrayList

class TopicService : JobService() {

    private val isNetworkAvailableAndConnected: Boolean
        get() {
            val cm = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            val isNetworkAvailable = cm.activeNetworkInfo != null
            return isNetworkAvailable && cm.activeNetworkInfo?.isConnected ?: false
        }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        return Service.START_STICKY
    }

    override fun onStartJob(jobParameters: JobParameters): Boolean {
        if (!isNetworkAvailableAndConnected) {
            Util.scheduleJob(this)
            return false
        }
        val topics = TopicList.getInstance(this)!!.topics
        for (t in topics) {
            if (t.enabled) {
                when (t) {
                    is YoutubeTopic -> {
                        val oldVideoIDs = t.youtubeTopicSearcher?.videoIds
                        t.youtubeTopicSearcher = YoutubeTopicSearcher(t)
                        t.youtubeTopicSearcher?.search(object : TopicCallback {
                            override fun onFinished() {
                                val newVideoIDs = t.youtubeTopicSearcher?.videoIds
                                val newVideoResults = t.youtubeTopicSearcher?.videoResults
                                val uniqueVideoIDs = ArrayList<String>()
                                for (i in (newVideoIDs?.indices ?: return)) {
                                    if (i < (newVideoResults?.size ?: return) &&
                                            !(oldVideoIDs?.contains(newVideoIDs[i]) ?: return) &&
                                            !t.previousNotifications.contains(newVideoIDs[i])) {
                                        uniqueVideoIDs.add(newVideoIDs[i])
                                        val title = "New Youtube Video from " + newVideoResults[i].snippet.channelTitle
                                        val body = newVideoResults[i].snippet.title
                                        t.previousNotifications.add(newVideoIDs[i])
                                        TopicList.getInstance(this@TopicService)?.updateTopic(t)
                                        Util.createYoutubeNotification(newVideoIDs[i], title, body, this@TopicService)
                                    }
                                }
                                Log.i(TAG, "YoutubeTopic refreshed")
                                Log.i(TAG, uniqueVideoIDs.toString())
                            }

                            override fun onStarted() {
                            }
                        })
                    }
                }
            }
        }
        Util.scheduleJob(this)
        return true
    }

    override fun onStopJob(jobParameters: JobParameters): Boolean {
        return true
    }

    companion object {
        const val TAG = "topic_service"
        const val YOUTUBE_CHANNEL_ID = "youtube_topic_channel"

        fun newIntent(context: Context): Intent {
            return Intent(context, TopicService::class.java)
        }
    }
}
